### PR TITLE
Apply offset after cast so that the right amount of bytes are skipped…

### PR DESCRIPTION
**Pull request by nwrkbiz**
…. (If offset is added to uin8_t* the wrong amount of bytes are skippet for concrete datatypes)

Use array operator as it is more readable / lower cognitive complexity, especially for array of strings.

### CODE REVIEW

File: src/koffi/src/ffi.cc [hunk 1]
```diff
@@ -1664,7 +1664,7 @@ static Napi::Value DecodeValue(const Napi::CallbackInfo &info)
 
     Napi::External<void> external = info[0].As<Napi::External<void>>();
     int64_t offset = has_offset ? info[1].As<Napi::Number>().Int64Value() : 0;
-    const uint8_t *ptr = (const uint8_t *)external.Data() + offset;
```
Comment (nwrkbiz) [inline src/koffi/src/ffi.cc:1667]: This does not work, you need to cast to the right type first. (Here the compiler does not know how many bytes to skip)

**Comment by Koromix** [comment]
I know, but I did it on purpose. I'm okay to re-evaluate though.

Right now I made this choice because:

- If you have a pointer to a struct and want to decode a single member in it, it's much easier with byte offsets.
- For arrays, you can simply multiply the offset with the size of the type, but the reverse is not really possible (at least with integer offsets).
- I've never liked how offsets work in C, I would prefer them to be byte-sized all the time. So personal preference.

But in the end, I think this `decode()` API is not intuitive enough, because it seems to do several stuff at once (decode one element in an object versus in an array). I have to think about it next week-end. Maybe we need to differentiate between offsets (byte-sized) and indices (element-sized), somehow?

**Comment by nwrkbiz** [comment]
O.k., I fully get your points (:

In my opinion the idea is good as there is the koffi.sizeof() function.

I will play around a little and maybe provide a PR/MR with some code examples for your documentation.

No I think it makes totally sense the way you described it. I just did not think about using this api for other purposes than accessing array values.

**Comment by Koromix** [comment]
Still, offset is byte sized but len depends on the type so it's not so clear cut I guess ^^

I've not yet solved this issue, but in the mean time decode() now supports record and array types.
It's not yet documented but you can see an example in the test suite: https://github.com/Koromix/rygel/blob/2e1ca1b46996792ef87a457dd8356d026ae3101c/src/koffi/test/callbacks.js#L189-L216

**Comment by nwrkbiz** [comment]
> Still, offset is byte sized but len depends on the type so it's not so clear cut I guess ^^

Hmm, then maybe instead of cunstructing a string a Unint8Array should be constructed. The caller than shall use a TextEncoder if he needs a string.

**Comment by Koromix** [comment]
I've kept the offset as-is (bytes), and released Koffi 2.2.

Here is the documentation for `koffi.decode()`: https://koffi.dev/callbacks#pointer-arguments

**Comment by nwrkbiz** [comment]
Top, thanks!

**Pull request closed without merging**